### PR TITLE
Expose commit hook

### PR DESCRIPTION
**Pull request by davidisaaclee**
### Checklist
- [x] I grant to recipients of this Project distribution a perpetual,
non-exclusive, royalty-free, irrevocable copyright license to reproduce, prepare
derivative works of, publicly display, sublicense, and distribute this
Contribution and such derivative works.
- [x] I certify that I am legally entitled to grant this license, and that this
Contribution contains no content requiring a license from any third party.

I made this PR largely by copying #210 and trusting tests. I just now switched a project to use my fork, and `commit_hook()` appears to be working as expected.

I think we could add the rollback hook very easily with this work, but I didn't need that. If you want that feature here, I'd be happy to implement it.

I committed a build made using Emscripten 3.1.70.

### CODE REVIEW

File: src/libadapters.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Method names for these signatures must be in src/asyncify_imports.json.
 const SIGNATURES = [
-  'ipp', // xProgress
+  'ipp', // xProgress, xCommitHook
```
Comment (davidisaaclee) [inline src/libadapters.js:3]: `xCommitHook` is made up: I couldn't find an official name for this callback. (I figured this is more helpful than harmful.)

File: src/libhook.js [hunk 2]
```diff
@@ -26,4 +26,26 @@
       });
     }
   };
-})();
\ No newline at end of file
+
+  Module['commit_hook'] = function(db, xCommitHook) {
```
Comment (rhashimoto) [inline src/libhook.js:30]: I think that using the same `pAsyncFlags` for both `update_hook()` and `commit_hook()` is a mistake here. I believe that calling `commit_hook()` will unset any previous `update_hook()` and vice versa.

I believe that using a separate IIFE for each function should work, as they will have their own `pAsyncFlags`.

**Review by rhashimoto — COMMENTED** [review]
Thanks for the PR! Found the commented issue on reading through your changes.

If you can please rebuild dist/ with EMSDK 3.1.61 that will make it easier for me to verify your binaries.

File: src/libadapters.js (hunk 1 above)
Comment (rhashimoto) [reply src/libadapters.js:3]: That's fine.

File: src/libhook.js (hunk 2 above)
Comment (davidisaaclee) [reply src/libhook.js:35]: good catch – [c5fd169](https://github.com/rhashimoto/wa-sqlite/pull/225/commits/c5fd16902b8fe013f1a857e8797fb689b63a78f3) adds a failing test which shows that you're right about this, and [7d68b39](https://github.com/rhashimoto/wa-sqlite/pull/225/commits/7d68b39e307a888acd866eeeafcfa522e06f5635) puts the commit hook in its own IIFE, which makes the test pass.

Comment (rhashimoto) [reply src/libhook.js:35]: Looks good! I just need to verify the binaries and then read through the code one final time.

File: test/callbacks.test.js [hunk 3]
```diff
@@ -422,4 +422,159 @@ for (const [key, factory] of FACTORIES) {
...
+      expect(rc).toEqual(SQLite.SQLITE_OK);
+
+      let updateHookInvocationsCount = 0;
+      sqlite3.update_hook(db, (...args) => {
+        updateHookInvocationsCount++;
+      });
+
+      let commitHookInvocationsCount = 0;
+      sqlite3.commit_hook(db, () => {
+        commitHookInvocationsCount++;
```
Comment (rhashimoto) [inline test/callbacks.test.js:568]: Minor nit: this commit callback should return a number, probably 0.

File: src/types/index.d.ts [hunk 4]
```diff
@@ -458,6 +458,19 @@ declare interface SQLiteAPI {
    */
   column_type(stmt: number, i: number): number;
 
+   /**
```
Comment (rhashimoto) [inline src/types/index.d.ts:461]: Minor nit: remove 1 space from indent before comment start.

File: src/types/index.d.ts [hunk 5]
```diff
@@ -458,6 +458,19 @@ declare interface SQLiteAPI {
...
+   * Register a commit hook
+   * 
+   * @see https://www.sqlite.org/c3ref/commit_hook.html
+   *
+   * @param db database pointer
+   * @param callback If a non-zero value is returned, commit is converted into
+   * a rollback; disables callback when null
+   */
+   commit_hook(
```
Comment (rhashimoto) [inline src/types/index.d.ts:470]: Minor nit: remove 1 space from indent before `commit_hook`.

**Review by rhashimoto — COMMENTED** [review]
I can build byte-for-byte identical binaries.

Just fix these cosmetic nits and I'll merge.

File: test/callbacks.test.js (hunk 3 above)
Comment (davidisaaclee) [reply test/callbacks.test.js:568]: good catch, [729af6f](https://github.com/rhashimoto/wa-sqlite/pull/225/commits/729af6f28c9e827fa6fc1cf7e6a0bdcad6feee69) - test still passes

File: src/types/index.d.ts (hunk 4 above)
Comment (davidisaaclee) [reply src/types/index.d.ts]: [dc9b8a1](https://github.com/rhashimoto/wa-sqlite/pull/225/commits/dc9b8a1bdfbd1eba443e1a652e1b75e84bacfd38)

File: src/types/index.d.ts (hunk 5 above)
Comment (davidisaaclee) [reply src/types/index.d.ts]: [dc9b8a1](https://github.com/rhashimoto/wa-sqlite/pull/225/commits/dc9b8a1bdfbd1eba443e1a652e1b75e84bacfd38)

**Pull request merged**